### PR TITLE
Implement Clock class

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
Another piece required for the block runner is a clock. Here I introduced the interface IClock and two concrete implementations `StdClock` and `YarpClock`. The former is useful for standalone applications while the latter becomes pivotal when you want to simulate your system with gazebo and using a shared clock. 

The clocks, both yarp, and std are implemented using singleton and they are instantiated using the factory paradigm. The `std` lof is the default clock. If you want to use the `yarp` one you should set the associated factory in the `ClockBuilder`.

Long story short. 
1. Do you want to use a normal std clock?
   ```cpp
   // this will return the time since the epoch
   auto now = BipedalLocomotion::clock().now();
   ```
2. Do you want to use yarp clock?
   ```cpp
   // Change the clock. Call it only once in your main. From now on you will use the yarp clock 
 
BipedalLocomotionnnn::System::ClockBuilder::setFactory(std::make_shared<BipedalLocomotion::System::YarpClockFactory>()));
    
   auto start = BipedalLocomotion::clock().now();
   ```

### CODE REVIEW

File: src/System/include/BipedalLocomotion/System/StdClock.h [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+ * auto end = BipedalLocomotion::clock().now();
+ * std::chrono::duration<double, std::milli> elapsed = end-start;
+ * \endcode
+ */
+class StdClock final : public IClock
+{
+public:
+    /**
+     * Get the system current time
+     * @return the current time since epoch computed with std::chrono::system_clock
```
Comment (S-Dafarra) [inline src/System/include/BipedalLocomotion/System/StdClock.h:38]:
```suggestion
     * @return the current time in milliseconds since epoch computed with std::chrono::system_clock
```

**Review by S-Dafarra — COMMENTED** [review]
Very nice! Some minor comments on the documentation.

File: src/System/include/BipedalLocomotion/System/StdClock.h (hunk 1 above)
Comment (traversaro) [reply src/System/include/BipedalLocomotion/System/StdClock.h:38]: Are you sure that are milliseconds?

Comment (S-Dafarra) [reply src/System/include/BipedalLocomotion/System/StdClock.h:38]: Not completely. That is what I assumed by reading https://github.com/dic-iit/bipedal-locomotion-framework/blob/914faefe0b1567bdddfe72c8b0dd178b09ec9190/src/System/include/BipedalLocomotion/System/StdClock.h#L30
but this is also the reason why I proposed the modification :grin:

Comment (S-Dafarra) [reply src/System/include/BipedalLocomotion/System/StdClock.h:38]: I think I might have misinterpreted indeed. It probably deserves an additional line of documentation

**Review by traversaro — APPROVED** [review]
Ok, but just to understand, how do you want to use this? If you want to use to run a periodic thread, it may be convenient to expose `sleep_for`/`sleep_until` ?

File: src/System/include/BipedalLocomotion/System/StdClock.h [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+#include <chrono>
+
+namespace BipedalLocomotion
+{
+namespace System
+{
+
+/**
+ * StdClock implements a the IClock interface using.
```
Comment (prashanthr05) [inline src/System/include/BipedalLocomotion/System/StdClock.h:21]:
```suggestion
 * StdClock implements the IClock interface using `<chrono>` from c++std library.
```

**Review by prashanthr05 — APPROVED** [review]
Minor changes.

**Comment by traversaro** [comment]
> yarp::os::Network::init(); //<------ This is required by yarp to use the clock. Call it only once in your main. 

I am bit dubious on this. Can't we just do that via RAII in the YARP Clock class?

**Comment by GiulioRomualdi** [comment]
> Can't we just do that via RAII in the YARP Clock class? Because if we call manually the init, we need also to call the fini.

Just to be sure, do you mean call `yarp::os::Network::ini()`  in the constructor of `YarpClock` and `fini` in the destruction? 

> Ok, but just to understand, how do you want to use this? If you want to use to run a periodic thread, it may be convenient to expose sleep_for/sleep_until ?

You are right! I cannot use `this_tread::sleep_for` when the yarp clock network is used!

**Comment by traversaro** [comment]
> > Can't we just do that via RAII in the YARP Clock class? Because if we call manually the init, we need also to call the fini.
> 
> Just to be sure, do you mean call `yarp::os::Network::ini()` in the constructor of `YarpClock` and `fini` in the destruction?

Either that, or just add a instance of the `yarp::os::Network` class that does that for you.

**Comment by GiulioRomualdi** [comment]
@traversaro here I implemented `sleepFor` and `sleepUntil` in  https://github.com/dic-iit/bipedal-locomotion-framework/pull/269/commits/2b3b9e7d83124ca9d6e881d51e7819234dcbc48d. Do you think it's fine?

**Review by traversaro — APPROVED** [review]

**Comment by traversaro** [comment]
> 
> 
> @traversaro I implemented `sleepFor` and `sleepUntil` in [2b3b9e7](https://github.com/dic-iit/bipedal-locomotion-framework/commit/2b3b9e7d83124ca9d6e881d51e7819234dcbc48d). Do you think it's fine?

Yes!

File: src/System/include/BipedalLocomotion/System/StdClock.h (hunk 1 above)
Comment (GiulioRomualdi) [reply src/System/include/BipedalLocomotion/System/StdClock.h:38]: Hi @S-Dafarra, from c++ documentation 
```cpp
template<    class Rep,    class Period = std::ratio<1>> 
class duration; 
```
It consists of a count of ticks of type Rep and a tick period, where the tick period is a compile-time rational constant representing the number of seconds from one tick to the next.
Notice that a `std::ratio` represents the tick period (i.e. the number of seconds per tick).
So if you write
```cpp
auto now = BipedalLocomotion::clock().now();
```
you will have a `chrono::duration` whose ticks are expressed in seconds. (i.e. if you print `now.count()` you will have the seconds from epoch)
However std::chrono::duration can be construct also with the following constructor
```cpp
template< class Rep2, class Period2 >
constexpr duration( const duration<Rep2,Period2>& d );
```
That is the one used in the example
```cpp
std::chrono::duration<double, std::milli> elapsed = end-start; 
```
Here `std::milli` is `Period2`.

In conclusion, the function returns a class whose number of seconds per tick is equal to 1. 
I can add the following line in the documentation
```
* @note `BipedalLocomoition::clock().now().count()` returns the number of seconds since epoch 
```

Comment (GiulioRomualdi) [reply src/System/include/BipedalLocomotion/System/StdClock.h:38]: @S-Dafarra let me know if you're fine with https://github.com/dic-iit/bipedal-locomotion-framework/pull/269/commits/005d7c158d327cef65a38c8d878e7ebf20818898

**Review by S-Dafarra — APPROVED** [review]

File: src/System/include/BipedalLocomotion/System/StdClock.h (hunk 1 above)
Comment (S-Dafarra) [reply src/System/include/BipedalLocomotion/System/StdClock.h:38]: I am! Thanks!
